### PR TITLE
Start muted if there are more than 3 participants. This closes #12.

### LINKED ISSUE #12: Start muted if there are more than 3 participants
Like it happens in Google Hangouts when you join a crowded session.

### PR DESCRIPTION
**Pull request by sandr3j**
Audio is now off by default and will be toggled if the number of obtained participants is less than 4 when the user joins.

### CODE REVIEW

File: src/app/components/room/room.service.js [hunk 1]
```diff
@@ -144,12 +144,21 @@
             ActionService.enterRoom(msg.id, username, connection);
             // Step 3. Establish WebRTC connection with the Janus server
             // Step 4a (parallel with 4b). Publish our feed on server
+
+            var participants = 0;
             connection.publish({
+              success: function() {
+                  //start session unmuted?
+                  if(participants < 4){
```
Comment (imobach) [inline src/app/components/room/room.service.js]: What about some kind of "variable" to hold that 4 with a descriptive name?

Comment (sandr3j) [inline src/app/components/room/room.service.js]: Sounds useful, if somebody doesn't want to use the default limit. I can add the variable at the beginning of doEnter().

Comment (imobach) [inline src/app/components/room/room.service.js]: On the other hand, I think that you can get the list of participants from `msg` here, right? So you won't need that `participants` variable.

Comment (sandr3j) [inline src/app/components/room/room.service.js]: I took this approach because msg.publishers instanceof Array could be false, so I have a default value.

Comment (imobach) [inline src/app/components/room/room.service.js]: Yes, but what I mean is that you can also check that is not `false` in that callback.

The (rare) problem is that if the callback is really fast and gets executed before the rest of the function, `participants` won't be set at all to the correct value. I'm not sure if such a thing can happen, but...

Comment (sandr3j) [inline src/app/components/room/room.service.js]: I had that thought but connection.publish is an async network operation and will never fire that fast, so I wanted to keep the code short.

**Comment by sandr3j** [comment]
based on your feedback I made some modifications

**Comment by ancorgs** [comment]
I like it in general so far. Have you tested it?

**Comment by sandr3j** [comment]
I test everything I commit ;)

**Comment by imobach** [comment]
Ok, so let's merge it. Thanks a lot for your contribution!

File: src/app/components/feed/connection-config.factory.js [hunk 2]
```diff
@@ -24,7 +24,7 @@
     return function(pluginHandle, wantedInit, jsep, ok) {
       var current = {};
       var requested = null;
-      var wanted = {audio: true, video: true};
+      var wanted = {audio: false, video: true};
```
Comment (ancorgs) [inline src/app/components/feed/connection-config.factory.js]: Sure you need to change this IN ADDITION to changing `feed-connection.factory.js`?

Comment (sandr3j) [inline src/app/components/feed/connection-config.factory.js]: It seems like it doesn't make a difference, but with audio: false I get "Connection configured Object {audio: false, video: true}" two times in the console, while with audio: true only one time. When I tested I got some configuration error and thought this was the cause but it isn't. So maybe I should revert this change..

**Comment by ancorgs** [comment]
> I test everything I commit ;)

That's what wanted to read ;)

File: src/app/components/feed/feed-connection.factory.js [hunk 3]
```diff
@@ -64,7 +64,7 @@
         options = options || {};
 
         var media = {videoRecv: false, audioRecv: false};
-        var cfg = {video: true, audio: true};
+        var cfg = {video: true, audio: false};
```
Comment (ancorgs) [inline src/app/components/feed/feed-connection.factory.js]: Even more, I probably wouldn't change this value directly here. I would add a new option to `publish`, similar to the already existing `noCamera`. Let's call it, for example, `muted`. And then I would use that argument when calling `publish` (see the other comment).

File: src/app/components/room/room.service.js [hunk 4]
```diff
@@ -144,7 +147,14 @@
             ActionService.enterRoom(msg.id, username, connection);
             // Step 3. Establish WebRTC connection with the Janus server
             // Step 4a (parallel with 4b). Publish our feed on server
+
             connection.publish({
+              success: function() {
```
Comment (ancorgs) [inline src/app/components/room/room.service.js]: instead of abusing `success` to activate the audio after having connected muted (quite tricky), I would use the new option explained in the other comment. Something like
```javascript
var startMuted = (msg.publishers instanceof Array) && msg.publishers.length > joinUnmutedLimit;
connection.publish({
    muted: startMuted,
     error: function() { connection.publish({noCamera: true, muted: startMuted}); }
});

File: src/app/components/room/room.service.js [hunk 5]
```diff
@@ -101,6 +101,9 @@
     function doEnter(username) {
       var $$rootScope = $rootScope;
       var connection = null;
+      
+      //max number of participants who will join session unmuted
+      var joinUnmutedLimit = 3;
```
Comment (ancorgs) [inline src/app/components/room/room.service.js]: IMO, this should be a configuration option in `config.json`. If set to null (or undef), it would mean "never start muted".

File: src/app/components/room/room.service.js (hunk 4 above)
Comment (ancorgs) [inline src/app/components/room/room.service.js]: It would be nice to explain the user what is happening. Showing a notification "you start muted because bla bla" would be a legit usage for the `success` callback (not saying it's the correct place to put it, but at least it would not be abusing it).

**Comment by ancorgs** [comment]
Now I can say I have done the real code review. I added several comments. To summarize:

- Don't touch `ConnectionConfig`, it's a low level layer that should not be affected.
- Add a new argument to `FeedConnection.publish` instead of changing the default behavior
- Use that new argument instead of abusing the `success` callback
- Move `joinUnmutedLimit` to `config.json`
- (Optional) a notification to the user, so they understand what's going on
